### PR TITLE
Track more classification metrics in training plans

### LINKED ISSUE #1958: Track more classification metrics in `SemiSupervisedTrainingPlan` and `ClassifierTrainingPlan`

### PR DESCRIPTION
**Pull request by martinkim0**
Closes #1958

### CODE REVIEW

File: scvi/train/_trainingplans.py [hunk 1]
```diff
@@ -708,6 +709,37 @@ def __init__(
...
+
+        accuracy = Accuracy(task="multiclass", num_classes=n_classes).to(logits.device)(
+            predicted_labels, true_labels
+        )
+        f1 = F1Score(task="multiclass", num_classes=n_classes).to(logits.device)(
+            predicted_labels, true_labels
+        )
+        auroc = AUROC(task="multiclass", num_classes=n_classes).to(logits.device)(
+            logits, true_labels.view(-1).long()
+        )
```
Comment (adamgayoso) [inline scvi/train/_trainingplans.py:736]: is there a way to avoid `.to`? In general this can be problematic on multi device training. I see it might be because you don't know n_classes a priori.

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/train/_trainingplans.py (hunk 1 above)
Comment (martinkim0) [reply scvi/train/_trainingplans.py:736]: I think it should be possible by moving the tensors to CPU before logging and then the metric will be on the same device by default

Comment (adamgayoso) [reply scvi/train/_trainingplans.py]: I though if you decalre the metrics in the init, lightning will handle all of this? Maybe we could detect n_classes in a different way

Comment (martinkim0) [reply scvi/train/_trainingplans.py]: Makes sense. n_classes can be passed into the training plan init from scanvi

File: scvi/train/_trainingplans.py [hunk 2]
```diff
@@ -1050,6 +1086,11 @@ def forward(self, *args, **kwargs):
         """Passthrough to the module's forward function."""
         return self.module(*args, **kwargs)
 
+    def compute_and_log_metrics(
+        self, loss_output: LossOutput, metrics: Dict[str, ElboMetric], mode: str
+    ):
+        assert loss_output.classification != {}
```
Comment (adamgayoso) [inline scvi/train/_trainingplans.py:1092]: let's raise a proper error here

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/module/base/_base_module.py [hunk 3]
```diff
@@ -74,6 +74,7 @@ class LossOutput:
     reconstruction_loss: LossRecord | None = None
     kl_local: LossRecord | None = None
     kl_global: LossRecord | None = None
+    classification: dict[str, Tensor] | None = field(default_factory=dict)
```
[the remark this answers is not in the document]
Comment (martinkim0) [reply scvi/module/base/_base_module.py:77]: Still not sure whether the keys in this argument should be strict or not. For now, I'm thinking that it should at least contain a classification loss, the ground truth labels, and either logits or predicted labels. logits are probably better since auroc does not accept predicted labels.

[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/module/base/_base_module.py:77]: my concern lies in the fact that all other arguments here are used in an explicit way, even `extra_metrics`, which just get logged. So I'm wondering if classification should also be a dataclass or more structured?

[the remark this answers is not in the document]
Comment (martinkim0) [reply scvi/module/base/_base_module.py:77]: I see your point, since all other args do get logged, but this isn't necessarily the case for `classification`. In that case, I feel like separating it out into a new dataclass would make sense

[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/module/base/_base_module.py:77]: alternatively, add labels and logits here just like kl_local and global

**Comment by martinkim0** [comment]
There's some repeated code here such as `log_with_mode`. I'm planning on a separate PR to refactor these more cleanly.

**Review by adamgayoso — APPROVED** [review]
